presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,250 @@
+        String requestBodyJson = planFragmentJsonCodec.toJson(planFragment);
+        final Request request = buildRequest(requestBodyJson);
+
+        try (Response response = httpClient.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                PlanConversionFailure failure = processResponseFailure(response);
+                throw new PrestoException(failure, "Error from native plan checker: " + failure.getMessage());
+            }
+        }
+        catch (final IOException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "I/O error getting native plan checker response", e);
+        }
+        finally {
+            LOG.debug("Native plan validation complete for plan fragment id: %s", planFragment.getId());
+        }
+    }
+
+    private Request buildRequest(String requestBodyJson)
+    {
+        Request.Builder builder = new Request.Builder()
+                .url(getPlanValidateUrl())
+                .addHeader("CONTENT_TYPE", "APPLICATION_JSON")
+                .post(RequestBody.create(JSON_CONTENT_TYPE, requestBodyJson));
+
+        return builder.build();
+    }
+
+    private PlanConversionFailure processResponseFailure(Response response) throws IOException
+    {
+        if (response.body() != null) {
+            PlanConversionResponse planConversionResponse = PLAN_CONVERSION_RESPONSE_JSON_CODEC.fromJson(response.body().bytes());
+            if (!planConversionResponse.getFailures().isEmpty()) {
+                return planConversionResponse.getFailures().get(0);
+            }
+            else {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "Error response without failure from native plan checker with code: " + response.code());
+            }
+        }
+        else {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Error response without body from native plan checker with code: " + response.code());
+        }
+    }
+
+    private static class CheckInternalVisitor
+            extends PlanVisitor<Boolean, Void>
+    {
+        @Override
+        public Boolean visitTableScan(TableScanNode tableScan, Void context)
+        {
+            TableHandle handle = tableScan.getTable();
+            return ConnectorId.isInternalSystemConnector(handle.getConnectorId());
+        }
+
+        @Override
+        public Boolean visitPlan(PlanNode node, Void context)
+        {
+            for (PlanNode child : node.getSources()) {
+                if (child.accept(this, context)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    public static class PlanConversionResponse
+    {
+        private final String status;
+        private final List<PlanConversionFailure> failures;
+
+        @JsonCreator
+        public PlanConversionResponse(
+                @JsonProperty("status") String status,
```

Comment:
Coming back to this, what does the code return that the HTTP status code doesn't give us?

Reply:
Sorry, didn't see the previous comment. Yes, it redundant. I only added so that there could be something in the response message to indicate if it was successful or not, but you could also infer that based on if there is a failure in the message or the HTTP status code.

I did figure this response message would change anyway once there is a need to get the converted plan back, but I'll go ahead and remove this status for now.